File: Difficulty: Medium/Indexes of Subarray Sum/indexes-of-subarray-sum.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

#include <vector>
using namespace std;

class Solution {
public:
    vector<int> subarraySum(vector<int> &arr, int target) {
        int n = arr.size();
        int start = 0, end = 0;
        int currentSum = 0;

        while (end < n) {
            // Add the current element to the sum
            currentSum += arr[end];

            // Shrink the window from the left if the sum exceeds the target
            while (currentSum > target && start <= end) {
                currentSum -= arr[start];
                start++;
            }

            // Check if the current sum equals the target
            if (currentSum == target) {
                return {start + 1, end + 1}; // 1-based indexing
            }

            // Expand the window
            end++;
        }

        // No subarray found
        return {-1};
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr;
        int d;
        string input;

        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }

        cin >> d;
        cin.ignore();

        Solution ob;
        vector<int> result = ob.subarraySum(arr, d);
        for (int i : result) {
            cout << i << " ";
        }
        cout << "\n~\n";
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Indexes of Subarray Sum/indexes-of-subarray-sum.java

class Solution {
    static ArrayList<Integer> subarraySum(int[] arr, int target) {
        ArrayList<Integer> result  = new ArrayList<>();
        int currSum=0;
        int start=0;
        for(int end =0; end<arr.length;end++){
            currSum +=arr[end];
            while (currSum>target && start<=end){
                currSum-=arr[start];
                start ++;
            }
            
            if(currSum ==target){
                result.add(start+1);
                result.add(end+1);
                return result;
            }
        }
        
        result.add(-1);
        return result;
    }
}
